### DISCUSSION TITLE
Hello, I am MichaelJachan. I would like to work with u on HED800 / 3G.

### DISCUSSION THREAD

**Discussion by MichaelJachan** [hed-standard/hed-specification#323] [Show and tell]
Dear HED Community! 

Ramon (I guess u know him) said, I can comment the code in here. Unfortunately (?), I have created a PullRequest (PR) with a very small code change (formulated as TODO-Comments)

If there is a way to annotate code (as if I were the PR-reviewer), I would like to use it! 

BR and BTW, I like HED, 

MJ:)

PS: I have reviewed the HED800.xml and i guess i have found 10-20 small issues in the codefile ... :)

**Comment by MichaelJachan** [comment]
I have some suggestions 4u, the authors (if i were a code reviewer) about the 800 XML file.

**Reply by MichaelJachan** [reply]
1) Please use a tool for xml code formatting. The indents shall be unified (ie, 2 or 4 spaces). it makes it easier in the end. 

Q: Which tool do u use to edit / view the original HED.develop.xml - file? I can recommend some IDE like VisualStudio ;)

**Reply by MichaelJachan** [reply]
2) Rename "Attribute/Sensory" to "Sensory-attribute". (This is the content of my PR, so reject my PR, please;)

**Reply by MichaelJachan** [reply]
3. Several Tags are missing STRING or NUMERICAL type attributes:

/Item/Object/Man-made-object/Device/IO-device/Input-device/Keyboard/Keyboard-key/#

/Item/Object/Man-made-object/Device/IO-device/Input-device/Keypad/Keypad-key/#

/Property/Agent-property/Agent-trait/Age/#

All Tags around: /Property/Data-property/Data-resolution/Printer-resolution/#

/Property/Data-property/Data-value/Spatiotemporal-value/Rate-of-change/Sampling-rate/#

/Property/Data-property/Data-value/Spatiotemporal-value/Rate-of-change/Refresh-rate/#

/Property/Data-property/Data-value/Spatiotemporal-value/Spatial-value/Angle/#

this and following: /Property/Informational-property/Metadata/CogAtlas/#

this and ff: /Property/Informational-property/Metadata/Pathname/#

**Reply by MichaelJachan** [reply]
4. "labelClass" might be forgotten to be deleted ...

**Reply by MichaelJachan** [reply]
5. I guess the xml-attribute "requireChild" is obsolete (or i do not understand its meaning, as it is used by just a few Tags ... )

**Reply by MichaelJachan** [reply]
6. I guess the xml-attribute "takesValue" is totally obsolete, the "#" gives all needed informations (afai need it)

**Reply by MichaelJachan** [reply]
7. Typo in Line 2754: <description>A numerical value betwee 0 and 1.</description>

**Reply by VisLab** [reply]
Hello Michael,
Thank you for your comments and suggestions.  

1.  I think you might be looking at a pre-release version of the schema.  The released 8.0.0 can be viewed at:
[https://www.hedtags.org/display_hed.html](https://www.hedtags.org/display_hed.html)

2. We never directly edit the XML.  Our current process is that we edit the mediawiki in hed-specification/hedwiki and use tools to generate. Most XML viewers ignore spaces and do their own spacing.  We typically use the viewer to look at for review.  We clearly need to document the process and work on improvements. **Along that line, I am going to add a PROPOSED.md which summarizes things that need to be changed in the next version and things that are under discussion.**   We can edit this as suggestions, reviews, issues, and discussions occur.  When there is agreement, we will move them into a test schema so the tools can test before release.  (We'll have to work on this process.)

3. We haven't yet taken a serious look at the value classes. We put the `valueClass` infrastructure in so that we could move from hard-coding accepted attributes to taking them directly from the schema.  The transition won't be instantaneous as it involves both validator/tools and changes to schema while maintaining support for previous versions.

4.  I have invited you to be a member. I will go through your suggests and add them to PROPOSED.m.  Future pull requests should make changes to PROPOSED.md.

**Reply by MichaelJachan** [reply]
So, your XML-source, which i can download in the end from https://github.com/hed-standard/hed-specification/blob/master/hedxml/HED8.0.0.xml, is generated by some Mediawiki-Software, i understand. 

Is there maybe another way for me to download the xml?

**Reply by MichaelJachan** [reply]
>     2\. We never directly edit the XML.  Our current process is that we edit the mediawiki in hed-specification/hedwiki and use tools to generate

Yes, these "tools to generate" is the place, where your programmer can care for a nice formatting, ie, indent in the xml.

**Reply by MichaelJachan** [reply]
8. IDEA: Would you like to consider providing separate numeric types for int and double (ie, float). 

For now, I am not able to distinguish btw. int and double, when I read your xml. 

Thank you :)

**Reply by MichaelJachan** [reply]
9. Why does the entry "/Property/Organizational-property/Event-context", which is the only entry with the "unique" attribute, does not have a # subtag?

**Reply by MichaelJachan** [reply]
>     4\. I have invited you to be a member. I will go through your suggests and add them to PROPOSED.m.  Future pull requests should make changes to PROPOSED.md.

Is it OK for you, if i keep on posting my comments here, instead of writing them into the PROPOSED file? I think, someone of your internal team shall take and preselect my issues from here and put them in there ... 


HOWEVER, I would like to become a reviewer of one of your pull-requests :)

**Reply by MichaelJachan** [reply]
10) Warnings in HED-MATLAB-TOOLS: 

WARNING: for plugin "HEDTools" version in the folder name "3.1.0" and in the eegplugin_ file "3.0.0" differ

EEGLAB: adding "HEDTools" v3.1.0 (see >> help eegplugin_hedtools) - new version 2.7.0 available

**Comment by VisLab** [comment]
This probably would be a good place to explain why the top-level schema changed between versions 8.0.0-beta.5 and 8.0.0.  

Before the release, we did an intensive review with three aims in mind:

1. Making sure that the current structure of the HED schema did not preclude overlaying a HED ontology for future linked data work.
2. Making sure that every term stood on its own --- that is if you saw the word without its hierarchy you would not misinterpret what it meant.
3. There weren't any terms in the schema that were obvious mistakes or in the wrong place.

**Results:**

- The review took longer than anticipated and ended up delaying the release of HED-3G to the beginning of August.

- The need to be compatible with a future ontology (an BFO-based OBO Foundry ontology, to be specific) triggered a top level reorganization. The main difficulty was that OBO has some requirements about relations, which were not satisfied in 8.0.0-beta. We ended up changing the top-level (`Agent-property`, `Data-Property`, `Task-property`, and `Attribute`) to (`Property`, and `Relation`).   
**Note:** As long as annotators use the recommended short-form, moving terms around doesn't affect the annotation, so this isn't as draconian as it might seem.

- There were some problematic individual terms that changed.  For example, `Left-side` has become `Left-side-of` to make it clearly a `Relation`. 
Another problematic tag was `Visual` which became `Visual-presentation`.  In general `Sensory` has been separated into subcategories `Sensory-attribute` and `Sensory-presentation` under `Sensory-property`.  This was probably the change that will affect previously annotated data the most, but it cleaned up a lot of clarity issues.

**Comment by VisLab** [comment]
The `requireChild` attribute is still under discussion.   The issue is whether `xxx` should always require a value if `xxx/#` is in the schema. 

An `xxx` tag that has a `#` child cannot be extended, even if its parent has an `extensionAllowed`. This restriction makes unambiguous downstream parsing possible. 

However, the issue is whether `xxx` should be required to have a value.  Our current thinking is to relax this requirement so that `xxx `can be grouped with tags from the `Data-value` sub-tree.  This issue has not been settled, so the attribute has been left in 8.0.0, but for the most part is not used.

**Reply by MichaelJachan** [reply]
(I have deleted my old answer-post, because it was nonsense. Now I understand requireChild much better)

Before i can fully understand the syntax of requireChild, i need to ask another, more basic, question. I have asked Ramon the same, already: 

Hed is a list of Tags, which can also be seen as a tree. Right? 
A Tree has a root node, from which many branching nodes originate. Nodes are under other nodes. In the end of each branch, a node without any child nodes is called a leaf. Right? 

==> So my Question is: Is any node from the tree a valid HedTag, OR are only the leave nodes valid HadTags?

**Reply by VisLab** [reply]
Every node is a valid HED tag.  There is no requirement that it be a leaf node. You can use any node and tools will convert to its full path to the root of its subtree.   The HED xml   has <HED> as the root of the hierarchy.  There is no HED in the tag path, that is just an XML thing.  Thus, the HED hierarchy is actually a set of trees with roots:  Event, Item, Action, Agent, Property, and Relation.

**Comment by VisLab** [comment]
A node with the tagGroup attribute means that it must appear in some tag
group (has parentheses around it somewhere). We are probably going to put
the Relation subtree in the category when we start doing more
sophisticated processing of relations.

A topLevelTagGroup means the node has to be in a tagGroup that is not
within other tag groups.  The four nodes that currently have this attribute
are: Definition, Event-context, Onset, Offset.  The consequence is that
these tag groups cannot be nested. For example, Definitions cannot be
nested.  This is because the Definition syntax is:

(Definition/defname, ( actual definition tags))

Right now topLevelTagGroup is being handled directly from validation using
the schema, but the part that there can't be any other tags besides the
Definition and its tag group is hard-coded in the validator.  In a future
round of schema revisions we will add another schema attribute to enforce
this, since there will be some other tags such as Event-context that fall
into this category.  Since we aren't handling anything about Event-context
yet except for its straight schema validation, we didn't do it for the
first round.  We don't want Onset and Offset in the same tag group either.
This is currently hard-coded, but I think it can be taken care of by a
single schema attribute.

**Comment by VisLab** [comment]
We have code in Python to do it from github.  What did you have in mind?

We have a bunch of web services which are accessible in matlab.  These are
available with examples of how to call them in the webtools/examples
subdirectory of hed-standard/hed-python on github.  All of these services
are callable from the server https://hedtools.ucsd.edu/hed.  If it would be
helpful  to have a web service to get the XML callable from MATLAB, let me
know.  I can easily put it in for the next deployment release of the tools.

**Comment by VisLab** [comment]
Yes, this has been a languishing issue for a while (#19).  Now that we are
getting around to value classes, I'll escalate it to the proposed so it
gets some attention.

**Comment by VisLab** [comment]
Could you post this as an issue on Github hed-standard/hed-matlab   so we
have a record of it being reported and fixed... Thanks!

https://github.com/hed-standard/hed-matlab/issues
